### DISCUSSION TITLE
Plugin API

### DISCUSSION THREAD

**Discussion by code-asher** [coder/code-server#4072] [Ideas]
code-server currently has a plugin API but it is intentionally not documented or supported.

Some questions for discussion, feel free to add more:

1. Should code-server even have a plugin API?
2. What sort of things should it be able to do? (Add endpoints, override endpoints, customize authentication, etc.)
3. Should we scrap the current plugin system and implement it as a VS Code extension instead? (Likely will require patching VS Code's extension API.)
4. Or should we re-implement the plugin API separately from VS Code?
5. Or should we keep the plugin API as-is and expand on it?
6. Should the plugin API be Express-based as it currently is?
7. What specifically should the plugin API look like?

**Comment by code-asher** [comment]
My thoughts:

1. Yes, it lowers the cost of adding features.
2. Add arbitrary middleware (unsure of security implications though) and add endpoints but not override existing ones.
3. Since we are moving closer to VS Code with the fork anyway I think it might make a lot more sense to implement as a built-in VS Code extension instead of using a bespoke plugin system even if it does mean maintaining a bigger patch.
4. No, see 3
5. No, see 3
6. It depends on whether the next form of code-server will be using Express or not. But since Express is convenient and a well-known package I think it might be wise to keep using it.
7. TBD

**Reply by code-asher** [reply]
Doing 3 means we would need our own `vscode` npm package for types. `@coder/vscode` perhaps.

**Reply by jsjoeio** [reply]
> I think it might make a lot more sense to implement as a built-in VS Code extension instead of using a bespoke plugin system even if it does mean maintaining a bigger patch.

I do like the sound of this because then we can build on top of the existing VS Code Extension infrastructure/API rather than inventing our own bespoke "plugin" system.

**Reply by bpmct** [reply]
How would a VS Code extension help with things such as auth at the middleware level? I'm sure it makes sense, I just don't understand

**Reply by jsjoeio** [reply]
> How would a VS Code extension help with things such as auth at the middleware level?

I can't tell you exactly how it works, but a [PR/feature](https://github.com/cdr/code-server/pull/4069) like this might be better suited coming in as a VS Code extension.

**Reply by code-asher** [reply]
Yeah I am not sure what the API should look like but here is a hasty example:

```
import * as vscode from "@coder/vscode"

function isValidCredentials(req) { ... }

export function activate() {
  // registerMiddleware would be something we add to the VS Code API.
  vscode.registerMiddleware((req, res, next) => {
    if (vscode.isAuthenticated()) {
      return next()
    }
    if (req.method === "post" && isValidCredentials(req)) {
      vscode.authenticate()
      return next()
    }
    req.write("my login page")
  })
}
```

But maybe middleware is bad way to do this and we should have a `vscode.registerAuthProvider` that specifically handles authentication use cases or something like that instead.

**Reply by code-asher** [reply]
I think there are auth mechanisms already built into VS Code as well so it could be worth checking into whether those fulfill this use case.

**Reply by bpmct** [reply]
Makes sense to me. @jsjoeio what are your thoughts on supporting these types of extensions within code-server?

**Reply by jsjoeio** [reply]
Makes sense to me! I think we should have some type of API like this so folks can build on top of code-server instead of having to fork and or submit PRs upstream.

We've had cases like this come up enough that I think it's time to make this a reality.

**Reply by jsjoeio** [reply]
> Since we are moving closer to VS Code with the fork anyway I think it might make a lot more sense to implement as a built-in VS Code extension instead of using a bespoke plugin system even if it does mean maintaining a bigger patch.

@code-asher if we go the VS Code extension route, what do the next steps look like?

**Reply by code-asher** [reply]
Next steps:

1. Look into any existing VS Code authentication code and see if it is anything like what we need or if we can tweak it to make it work for us.
2. If that is a bust then put up a proposal for the required API changes.

**Reply by code-asher** [reply]
I should also mention that I think the first will only work once we fully convert to the fork. Right now we do not even spawn VS Code until we need it so we cannot rely on it for authentication. So we may need to punt on actual execution until that work is completed.

**Reply by jsjoeio** [reply]
> Right now we do not even spawn VS Code until we need it so we cannot rely on it for authentication. So we may need to punt on actual execution until that work is completed.

Ahh good point. Okay we'll reconsider this once we convert to the fork.

**Reply by sdissegna-maystreet** [reply]
A problem I see with using vscode extensions for auth, is that extensions are activated on each user connection, and only after the main workbench is already working and a websocket connection has already been established. At this point it's way too late to block unauthorized users, as vscode is already usable while extensions are loading. An unauthorized user could have already opened a terminal before the extension is ready. If the user is not authorized, you'd then have to forcefully close the active websocket.

A separate auth system would be a lot simpler. vscode extensions are designed to be controlled by the main workbench (it's the workbench that decides if an extension should even run, and when), while you'd want authentication to be the one in control of the rest of the system.

**Reply by code-asher** [reply]
That is a really, really good point!

Looks like we will need to preserve code-server's own plugin system and implement auth that way. Or we could scrap the plugin system and only support authentication modules.
